js/style/style_layer.js

```diff
@@ -59,21 +66,56 @@ StyleLayer.prototype = {
     resolvePaint: function() {
         for (var key in this._layer) {
             var match = key.match(/^paint(?:\.(.*))?$/);
-            if (!match) continue;
-            this._classes[match[1] || ''] = new StyleDeclarationSet('paint', this.type, this._layer[key]);
+            if (match) {
+                var klass = match[1] || '';
+                for (var name in this._layer[key]) {
+                    this.setPaintProperty(name, this._layer[key][name], klass);
+                }
+            }
         }
     },
 
     setPaintProperty: function(name, value, klass) {
-        var declarations = this._classes[klass || ''];
-        if (!declarations) {
-            declarations = this._classes[klass || ''] = new StyleDeclarationSet('paint', this.type, {});
+        if (endsWith(name, TRANSITION_SUFFIX)) {
+            if (!this._classPaintTransitions[klass || '']) {
+                this._classPaintTransitions[klass || ''] = {};
+            }
+            this._classPaintTransitions[klass || ''][name] = value;
+        } else {
+            if (!this._classPaintDeclarations[klass || '']) {
+                this._classPaintDeclarations[klass || ''] = {};
+            }
+            this._classPaintDeclarations[klass || ''][name] = new StyleDeclaration(this._paintSpecifications[name], value);
         }
-        declarations[name] = value;
     },
 
     getPaintProperty: function(name, klass) {
-        return this._classes[klass || ''] && this._classes[klass || ''][name];
+        klass = klass || '';
+        if (endsWith(name, TRANSITION_SUFFIX)) {
+            return (
+                this._classPaintTransitions[klass] &&
+                this._classPaintTransitions[klass][name]
+            );
+        } else {
+            return (
+                this._classPaintDeclarations[klass] &&
+                this._classPaintDeclarations[klass][name] &&
+                this._classPaintDeclarations[klass][name].value
+            );
```

Comment:
As with above, one branch is operating on `_classPaint`**`Declarations`** and one on `_classPaint`**`Transitions`**. I think those property names are too long. Shortened in the next commit.